test/e2e/ftccontroller.go

```diff
@@ -0,0 +1,195 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pborman/uuid"
+
+	apiextv1b1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	apiextv1b1client "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/typed/apiextensions/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/rest"
+
+	"sigs.k8s.io/kubefed/pkg/apis/core/typeconfig"
+	fedv1b1 "sigs.k8s.io/kubefed/pkg/apis/core/v1beta1"
+	genericclient "sigs.k8s.io/kubefed/pkg/client/generic"
+	"sigs.k8s.io/kubefed/pkg/controller/util"
+	"sigs.k8s.io/kubefed/pkg/kubefedctl"
+	kfenable "sigs.k8s.io/kubefed/pkg/kubefedctl/enable"
+	kfenableopts "sigs.k8s.io/kubefed/pkg/kubefedctl/options"
+	"sigs.k8s.io/kubefed/test/common"
+	"sigs.k8s.io/kubefed/test/e2e/framework"
+
+	. "github.com/onsi/ginkgo"
+)
+
+var _ = Describe("FTC controller", func() {
+	f := framework.NewKubeFedFramework("ftc-controller")
+	tl := f.Logger()
+
+	var client genericclient.Client
+
+	BeforeEach(func() {
+		if framework.TestContext.RunControllers() {
+			var controllerFixture *framework.ControllerFixture
+			controllerFixture, _ = framework.NewFederatedTypeConfigControllerFixture(tl, f.ControllerConfig())
+			f.RegisterFixture(controllerFixture)
+			client = f.Client("ftc-controller")
+		}
+	})
+
+	It("should be refreshed if the target CRD version changed for a federated CRD", func() {
+		if !framework.TestContext.RunControllers() {
+			framework.Skipf("FTC controller can only be tested when controllers are running in-process.")
```

Comment:
FTC controller test skipped for non-in-memory case.
Any problem with it?

Reply:
I don't think there's a problem with it. But it seems contrary to the default choice of testing against the deployed controllers instead of in-memory only. Is there a reason to use in-memory only since that's primarily for development purposes? I feel as though we should be prioritizing testing against the main deployment method if possible.

Reply:
Messing with FTC controller state of the deployed controller could have unintended consequences, so testing in-memory ensures that the controller can be isolated to a test namespace.